kmath-core/src/commonMain/kotlin/space/kscience/kmath/operations/Algebra.kt

```diff
@@ -247,11 +248,31 @@ public interface RingOperations<T> : GroupOperations<T> {
  */
 public interface Ring<T> : Group<T>, RingOperations<T> {
     /**
-     * neutral operation for multiplication
+     * The neutral element of multiplication
      */
     public val one: T
 }
 
+@UnstableKMathAPI
+public fun <T> Ring<T>.pow(base: T, exponent: ULong): T = when {
+    this == zero && exponent > 0UL -> zero
+    this == one -> base
+    this == -one -> powWithoutOptimization(base, exponent % 2UL)
+    else -> powWithoutOptimization(base, exponent)
+}
+
+@UnstableKMathAPI
+public fun <T> Ring<T>.pow(base: T, exponent: UInt): T = pow(base, exponent.toULong())
+
+private fun <T> Ring<T>.powWithoutOptimization(base: T, exponent: ULong): T = when (exponent) {
```

Comment:
Move pow operation to `algebraExtensions.kt`. It actually already have those functions: https://github.com/mipt-npm/kmath/blob/ca02f5406d06d18d07909f3d48508e1a99701fa5/kmath-core/src/commonMain/kotlin/space/kscience/kmath/operations/algebraExtensions.kt#L107, so you shoud probably just replace the implementation and use UInt instead of Int.

Reply:
Why not ULong?

Reply:
Also, I see a bug with Field<T>.power:
```kotlin
public fun <T> Field<T>.power(arg: T, power: Int): T {
    require(power != 0 || arg != zero) { "The $zero raised to $power is not defined." }
    if (power == 0) return one
    if (power < 0) return one / (this as Ring<T>).power(arg, -power)
    return (this as Ring<T>).power(arg, power)
}
```
If [power] is Int.MIN_VALUE, this will fail

Reply:
The implementation could be wrong in corner cases. It is not properly tested. As for ULong, the major bulk of operations on JVM and in kotlin use Int. I can't see a case where people would use more than Int for a power (please show them if you know them) and using Ulong would force users to do an additional unnecessary conversion.

Reply:
I am absolutely sure that there is no need to power numbers by ULong ranged numbers, however, this may be ok for some other Rings (albeit I don't know such ones).

Reply:
Also, 0^0 is not defined here while it is 1 in java's BigInt. Which convention to choose?
```kotlin
public fun <T> Ring<T>.power(arg: T, power: Int): T {
    require(power >= 0) { "The power can't be negative." }
    require(power != 0 || arg != zero) { "The $zero raised to $power is not defined." }
    if (power == 0) return one
    var res = arg
    repeat(power - 1) { res *= arg }
    return res
}
```

Reply:
Let's stick to UInt for now because I still can't imagine a case for longs. Since we are using extensions, it is easy to add another one if somebody needs it. As for 0^0, please fix it. I don't remember where this method comes from, but it was quite obviously forgotten.

Reply:
Do you mean that 0^0 is expected to be 1?

Reply:
*I published the changes.*